manpages/fsck.fat.8.in
.\" fsck.fat.8 - manpage for fsck.fat
.\"
.\" Copyright (C) 2006-2014 Daniel Baumann <daniel@debian.org>
.\"
.\" This program is free software: you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation, either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program. If not, see <http://www.gnu.org/licenses/>.
.\"
.\" The complete text of the GNU General Public License
.\" can be found in /usr/share/common-licenses/GPL-3 file.
.\"
.\"
.TH FSCK.FAT 8 2017-06-28 "dosfstools @PACKAGE_VERSION@"
.SH NAME
fsck.fat \- check and repair MS-DOS FAT filesystems
.\" ----------------------------------------------------------------------------
.SH SYNOPSIS
\fBfsck.fat\fP [\fIOPTIONS\fP] \fIDEVICE\fP
.\" ----------------------------------------------------------------------------
.SH DESCRIPTION
\fBfsck.fat\fP verifies the consistency of MS-DOS filesystems and optionally
tries to repair them.
.PP
The following filesystem problems can be corrected (in this order):
.IP "\(bu" 4
FAT contains invalid cluster numbers.
Cluster is changed to EOF.
.IP "\(bu" 4
File's cluster chain contains a loop.
The loop is broken.
.IP "\(bu" 4
Bad clusters (read errors).
The clusters are marked bad and they are removed from files owning them.
This check is optional.
.IP "\(bu" 4
Directories with a large number of bad entries (probably corrupt).
The directory can be deleted.
.IP "\(bu" 4
Files . and .. are non-directories.
They can be deleted or renamed.
.IP "\(bu" 4
Directories . and .. in root directory.
They are deleted.
.IP "\(bu" 4
Bad filenames.
They can be renamed.
.IP "\(bu" 4
Duplicate directory entries.
They can be deleted or renamed.
.IP "\(bu" 4
Directories with non-zero size field.
Size is set to zero.
.IP "\(bu" 4
Directory . does not point to parent directory.
The start pointer is adjusted.
.IP "\(bu" 4
Directory .. does not point to parent of parent directory.
The start pointer is adjusted.
.IP "\(bu" 4
Start cluster number of a file is invalid.
The file is truncated.
.IP "\(bu" 4
File contains bad or free clusters.
The file is truncated.
.IP "\(bu" 4
File's cluster chain is longer than indicated by the size fields.
The file is truncated.
.IP "\(bu" 4
Two or more files share the same cluster(s).
All but one of the files are truncated.
If the file being truncated is a directory file that has already been read, the
filesystem check is restarted after truncation.
.IP "\(bu" 4
File's cluster chain is shorter than indicated by the size fields.
The file is truncated.
.IP "\(bu" 4
Volume label in root directory or label in boot sector is invalid.
Invalid labels are removed.
.IP "\(bu" 4
Volume label in root directory and label in boot sector are different.
Volume label from root directory is copied to boot sector.
.IP "\(bu" 4
Clusters are marked as used but are not owned by a file.
They are marked as free.
.PP
Additionally, the following problems are detected, but not repaired:
.IP "\(bu" 4
Invalid parameters in boot sector
.IP "\(bu" 4
Absence of or too many . and .. entries in non-root directories
.IP "\(bu" 4
\&. and .. in a non-root directories are not the very first two entries
.PP
When \fBfsck.fat\fP checks a filesystem, it accumulates all changes in memory
and performs them only after all checks are complete.
This can be disabled with the \fB\-w\fP option.
.PP
Two different variants of the FAT filesystem are supported.
Standard is the FAT12, FAT16 and FAT32 filesystems as defined by Microsoft and
widely used on hard disks and removable media like USB sticks and SD cards.
The other is the legacy Atari variant used on Atari ST.
.PP
There are some minor differences in Atari format:
Some boot sector fields are interpreted slightly different, and the special FAT
entries for end-of-file and bad cluster can be different.
Under MS-DOS 0xfff8 is used for EOF and Atari employs 0xffff by default, but
both systems recognize all values from 0xfff8\(en0xffff as end-of-file.
MS-DOS uses only 0xfff7 for bad clusters, where on Atari values 0xfff0\(en0xfff7
are for this purpose (but the standard value is still 0xfff7).
.\" ----------------------------------------------------------------------------
.SH OPTIONS
.IP "\fB\-a\fP" 4
Automatically repair the filesystem.
No user intervention is necessary.
Whenever there is more than one method to solve a problem, the least
destructive approach is used.
.IP "\fB\-A\fP" 4
Select using the Atari variation of the FAT filesystem if that isn't active
already, otherwise select standard FAT filesystem.
This is selected by default if \fBmkfs.fat\fP is run on 68k Atari Linux.
.IP "\fB-b\fP" 4
Make read-only boot sector check.
.IP "\fB-c\fP \fIPAGE\fP" 4
Use DOS codepage \fIPAGE\fP to decode short file names.
By default codepage 850 is used.
.IP "\fB\-d\fP \fIPATH\fP" 4
Delete the specified file.
If more than one file with that name exist, the first one is deleted.
This option can be given more than once.
.IP "\fB\-f\fP" 4
Salvage unused cluster chains to files.
By default, unused clusters are added to the free disk space except in auto mode
(\fB\-a\fP).
.IP "\fB\-l\fP" 4
List path names of files being processed.
.IP "\fB\-n\fP" 4
No-operation mode: non-interactively check for errors, but don't write
anything to the filesystem.
.IP "\fB\-p\fP" 4
Same as \fB\-a\fP, for compatibility with other *fsck.
.IP "\fB\-r\fP" 4
Interactively repair the filesystem.
The user is asked for advice whenever there is more than one approach to fix an
inconsistency.
This is the default mode and the option is only retained for backwards
compatibility.
.IP "\fB\-S\fP" 4
Consider short (8.3) file names with spaces in the middle to be invalid, like
previous versions of this program did.
While such file names are not forbidden by the FAT specification, and were never
treated as errors by Microsoft file system checking tools, many DOS programs are
unable to handle files with such names. Using this option can make them
accessible to these programs.
.IP "" 4
Short file names which \fIstart\fP with a space are considered invalid
regardless of this option's setting.
.IP "" 4
Previous versions of this program exceptionally treated \fIEA DATA . SF\fP and
\fIWP ROOT . SF\fP as valid short names; using this option does not preserve
that exception.
.IP "\fB\-t\fP" 4
Mark unreadable clusters as bad.
.IP "\fB\-u\fP \fIPATH\fP" 4
Try to undelete the specified file.
\fBfsck.fat\fP tries to allocate a chain of contiguous unallocated clusters
beginning with the start cluster of the undeleted file.
This option can be given more than once.
.IP "\fB\-v\fP" 4
Verbose mode.
Generates slightly more output.
.IP "\fB\-V\fP" 4
Perform a verification pass.
The filesystem check is repeated after the first run.
The second pass should never report any fixable errors.
It may take considerably longer than the first pass, because the first pass may
have generated long list of modifications that have to be scanned for each disk
read.
.IP "\fB\-\-variant\fP \fITYPE\fP" 4
Create a filesystem of variant \fITYPE\fP.
Acceptable values are \fIstandard\fP and \fIatari\fP (in any combination of
upper/lower case).
See above under DESCRIPTION for the differences.
.IP "\fB\-w\fP" 4
Write changes to disk immediately.
.IP "\fB\-y\fP" 4
Same as \fB\-a\fP (automatically repair filesystem) for compatibility with other
fsck tools.
.IP "\fB\-\-help\fP" 4
Display help message describing usage and options then exit.
.\" ----------------------------------------------------------------------------
.SH "EXIT STATUS"
.IP "0" 4
No recoverable errors have been detected.
.IP "1" 4
Recoverable errors have been detected or \fBfsck.fat\fP has discovered an
internal inconsistency.
.IP "2" 4
Usage error.
\fBfsck.fat\fP did not access the filesystem.
.\" ----------------------------------------------------------------------------
.SH FILES
.IP "\fIfsck0000.rec\fP, \fIfsck0001.rec\fP, ..." 4
When recovering from a corrupted filesystem, \fBfsck.fat\fP dumps recovered data
into files named \fIfsckNNNN.rec\fP in the top level directory of the filesystem.
.\" ----------------------------------------------------------------------------
.SH BUGS
.IP "\(bu" 4
Does not create . and .. files where necessary.
.IP "\(bu" 4
Does not remove entirely empty directories.
.IP "\(bu" 4
Should give more diagnostic messages.
.IP "\(bu" 4
Undeleting files should use a more sophisticated algorithm.
.\" ----------------------------------------------------------------------------
.SH SEE ALSO
.BR fatlabel (8),
.BR mkfs.fat (8)
.\" ----------------------------------------------------------------------------
.SH HOMEPAGE
The home for the \fBdosfstools\fP project is its
.UR https://github.com/dosfstools/dosfstools
GitHub project page
.UE .
.\" ----------------------------------------------------------------------------
.SH AUTHORS
\fBdosfstools\fP were written by
.MT werner.almesberger@\:lrc.di.epfl.ch
Werner Almesberger
.ME ,
.MT Roman.Hodek@\:informatik.\:uni-erlangen.de
Roman Hodek
.ME ,
and others.
The current maintainer is
.MT aeb@\:debian.org
Andreas Bombe
.ME .


src/fsck.fat.c
/* fsck.fat.c - User interface

   Copyright (C) 1993 Werner Almesberger <werner.almesberger@lrc.di.epfl.ch>
   Copyright (C) 1998 Roman Hodek <Roman.Hodek@informatik.uni-erlangen.de>
   Copyright (C) 2008-2014 Daniel Baumann <mail@daniel-baumann.ch>

   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program. If not, see <http://www.gnu.org/licenses/>.

   The complete text of the GNU General Public License
   can be found in /usr/share/common-licenses/GPL-3 file.
*/

/* FAT32, VFAT, Atari format support, and various fixes additions May 1998
 * by Roman Hodek <Roman.Hodek@informatik.uni-erlangen.de> */

#include "version.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <termios.h>
#include <getopt.h>

#include "common.h"
#include "fsck.fat.h"
#include "io.h"
#include "boot.h"
#include "fat.h"
#include "file.h"
#include "check.h"
#include "charconv.h"

int rw = 0, list = 0, test = 0, verbose = 0;
int no_spaces_in_sfns = 0;
int boot_only = 0;
unsigned n_files = 0;
void *mem_queue = NULL;

static struct termios original_termios;


static void restore_termios(void)
{
    tcsetattr(0, TCSAFLUSH, &original_termios);
}


static void usage(char *name, int exitval)
{
    fprintf(stderr, "Usage: %s [OPTIONS] DEVICE\n", name);
    fprintf(stderr, "Check FAT filesystem on DEVICE for errors.\n");
    fprintf(stderr, "\n");
    fprintf(stderr, "Options:\n");
    fprintf(stderr, "  -a              automatically repair the filesystem\n");
    fprintf(stderr, "  -A              toggle Atari variant of the FAT filesystem\n");
    fprintf(stderr, "  -b              make read-only boot sector check\n");
    fprintf(stderr, "  -c N            use DOS codepage N to decode short file names (default: %d)\n",
	    DEFAULT_DOS_CODEPAGE);
    fprintf(stderr, "  -d PATH         drop file with name PATH (can be given multiple times)\n");
    fprintf(stderr, "  -f              salvage unused chains to files\n");
    fprintf(stderr, "  -l              list path names\n");
    fprintf(stderr, "  -n              no-op, check non-interactively without changing\n");
    fprintf(stderr, "  -p              same as -a, for compat with other *fsck\n");
    fprintf(stderr, "  -r              interactively repair the filesystem (default)\n");
    fprintf(stderr, "  -S              disallow spaces in the middle of short file names\n");
    fprintf(stderr, "  -t              test for bad clusters\n");
    fprintf(stderr, "  -u PATH         try to undelete (non-directory) file that was named PATH (can be\n");
    fprintf(stderr, "                    given multiple times)\n");
    fprintf(stderr, "  -v              verbose mode\n");
    fprintf(stderr, "  -V              perform a verification pass\n");
    fprintf(stderr, "  --variant=TYPE  handle variant TYPE of the filesystem\n");
    fprintf(stderr, "  -w              write changes to disk immediately\n");
    fprintf(stderr, "  -y              same as -a, for compat with other *fsck\n");
    fprintf(stderr, "  --help          print this message\n");
    exit(exitval);
}

int main(int argc, char **argv)
{
    DOS_FS fs;
    int salvage_files, verify, c;
    uint32_t free_clusters = 0;
    struct termios tio;

    enum {OPT_HELP=1000, OPT_VARIANT};
    const struct option long_options[] = {
	    {"variant", required_argument, NULL, OPT_VARIANT},
	    {"help",    no_argument,       NULL, OPT_HELP},
	    {0,}
    };

    if (!tcgetattr(0, &original_termios)) {
	tio = original_termios;
	tio.c_lflag &= ~(ICANON | ECHO);
	tcsetattr(0, TCSAFLUSH, &tio);
	atexit(restore_termios);
    }

    memset(&fs, 0, sizeof(fs));
    salvage_files = verify = 0;
    rw = interactive = 1;
    check_atari();

    while ((c = getopt_long(argc, argv, "Aac:d:bflnprStu:vVwy",
				    long_options, NULL)) != -1)
	switch (c) {
	case 'A':		/* toggle Atari format */
	    atari_format = !atari_format;
	    break;
	case 'a':
	case 'p':
	case 'y':
	    rw = 1;
	    interactive = 0;
	    salvage_files = 1;
	    break;
	case 'b':
	    rw = 0;
	    interactive = 0;
	    boot_only = 1;
	    break;
	case 'c':
	    set_dos_codepage(atoi(optarg));
	    break;
	case 'd':
	    file_add(optarg, fdt_drop);
	    break;
	case 'f':
	    salvage_files = 1;
	    break;
	case 'l':
	    list = 1;
	    break;
	case 'n':
	    rw = 0;
	    interactive = 0;
	    break;
	case 'r':
	    rw = 1;
	    interactive = 1;
	    break;
	case 'S':
	    no_spaces_in_sfns = 1;
	    break;
	case 't':
	    test = 1;
	    break;
	case 'u':
	    file_add(optarg, fdt_undelete);
	    break;
	case 'v':
	    verbose = 1;
	    break;
	case 'V':
	    verify = 1;
	    break;
	case OPT_VARIANT:
	    if (!strcasecmp(optarg, "standard")) {
		    atari_format = 0;
	    } else if (!strcasecmp(optarg, "atari")) {
		    atari_format = 1;
	    } else {
		    fprintf(stderr, "Unknown variant: %s\n", optarg);
		    usage(argv[0], 2);
	    }
	    break;
	case 'w':
	    write_immed = 1;
	    break;
	case OPT_HELP:
	    usage(argv[0], 0);
	    break;
	case '?':
	    usage(argv[0], 2);
	    break;
	default:
	    fprintf(stderr,
		    "Internal error: getopt_long() returned unexpected value %d\n", c);
	    exit(3);
	}
    set_dos_codepage(-1);	/* set default codepage if none was given in command line */
    if ((test || write_immed) && !rw) {
	fprintf(stderr, "-t and -w can not be used in read only mode\n");
	exit(2);
    }
    if (optind != argc - 1)
	usage(argv[0], 2);

    printf("fsck.fat " VERSION " (" VERSION_DATE ")\n");
    fs_open(argv[optind], rw);

    read_boot(&fs);
    if (boot_only)
	goto exit;

    if (verify)
	printf("Starting check/repair pass.\n");
    while (read_fat(&fs), scan_root(&fs))
	qfree(&mem_queue);
    check_label(&fs);
    if (test)
	fix_bad(&fs);
    if (salvage_files)
	reclaim_file(&fs);
    else
	reclaim_free(&fs);
    if (!atari_format)
	check_dirty_bits(&fs);
    free_clusters = update_free(&fs);
    file_unused();
    qfree(&mem_queue);
    if (verify) {
	n_files = 0;
	printf("Starting verification pass.\n");
	read_fat(&fs);
	scan_root(&fs);
	check_label(&fs);
	reclaim_free(&fs);
	if (!atari_format)
	    check_dirty_bits(&fs);
	qfree(&mem_queue);
    }
    release_fat(&fs);

exit:
    if (!write_immed && fs_changed()) {
	if (rw) {
	    printf("\n*** Filesystem was changed ***\n");
	    if (interactive)
		printf("The changes have not yet been written, you can still choose to leave the\n"
		       "filesystem unmodified:\n");

	    rw = get_choice(1, "Writing changes.",
			    2,
			    1, "Write changes",
			    2, "Leave filesystem unchanged") == 1;
	} else
	    printf("\nLeaving filesystem unchanged.\n");
    }

    if (!boot_only)
	printf("%s: %u files, %lu/%lu clusters\n", argv[optind],
	       n_files, (unsigned long)fs.data_clusters - free_clusters,
	       (unsigned long)fs.data_clusters);

    return fs_close(rw) ? 1 : 0;
}
